### PR TITLE
fix: Make ResourceUtils.getNumPlurals more consistent

### PR DESCRIPTION
**Pull request by seanf**
We should always fall back on locale defaults, if available, whether
the plurals header is absent, empty or unparseable.

### CODE REVIEW

**Comment by seanf** [comment]
@zanata-jenkins test this please

**Comment by seanf** [comment]
@carlosmunoz I could use a hand debugging the 404s which happen when the tests try to log start/stop (commit a6fdcca).

**Comment by seanf** [comment]
> @carlosmunoz I could use a hand debugging the 404s which happen when the tests try to log start/stop (commit a6fdcca).

Never mind: I didn't realise I had to register each endpoint class in zanata-test-war's JAX-RS Application class.

File: zanata-war/src/main/java/org/zanata/rest/service/ResourceUtils.java [hunk 1]
```diff
@@ -1479,4 +1489,11 @@ public static void validateExtensions(Set<String> requestedExt) {
         }
     }
 
+    public static class PluralParseException extends RuntimeException {
```
Comment (aeng) [inline zanata-war/src/main/java/org/zanata/rest/service/ResourceUtils.java]: Shouldn't this be a class itself?

File: zanata-war/src/main/java/org/zanata/rest/service/ResourceUtils.java [hunk 2]
```diff
@@ -108,6 +109,7 @@
     private static final String CONTENT_TYPE_HDR = HeaderFields.KEY_ContentType;
     private static final String PLURAL_FORMS_HDR = "Plural-Forms";
 
+    // TODO we should need only one regex, with a capture group for nplurals
```
Comment (aeng) [inline zanata-war/src/main/java/org/zanata/rest/service/ResourceUtils.java]: Isn't this only 1 pattern?

Comment (seanf) [inline zanata-war/src/main/java/org/zanata/rest/service/ResourceUtils.java]: There are three patterns in this file, but they didn't change in this PR, so they're not entirely visible here.

File: zanata-war/src/main/java/org/zanata/rest/service/ResourceUtils.java (hunk 1 above)
Comment (seanf) [inline zanata-war/src/main/java/org/zanata/rest/service/ResourceUtils.java]: It is, it just lives inside the ResourceUtils namespace.

I could make it a top-level class, but I don't think it's needed.

**Comment by aeng** [comment]
:+1:

**Comment by seanf** [comment]
@zanata-jenkins test this please

**Comment by seanf** [comment]
@zanata-jenkins test this please

**Comment by seanf** [comment]
@zanata-jenkins test this please

**Comment by djansen-redhat** [comment]
@seanf uploading the files attached (with broken nplurals in the .po header) results in a plurals of 1 for Arabic, where I expected 6.  Could you clarify this for me?
[DJansenTest.zip](https://github.com/zanata/zanata-server/files/183621/DJansenTest.zip)

**Comment by seanf** [comment]
@djansen-redhat I don't suppose you have the server log from that test?

On 22 March 2016 at 16:49, Damian Jansen <notifications@github.com> wrote:

> @seanf <https://github.com/seanf> uploading the files attached (with
> broken nplurals in the .po header) results in a plurals of 1 for Arabic,
> where I expected 6. Could you clarify this for me?
> DJansenTest.zip
> <https://github.com/zanata/zanata-server/files/183621/DJansenTest.zip>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zanata/zanata-server/pull/958#issuecomment-199665662>
>



-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat

**Comment by seanf** [comment]
@djansen-redhat the zanata.xml in that zip had a URL with `https` together with port `8080`. Did you really have Zanata running HTTPS on port 8080?

**Comment by seanf** [comment]
@djansen-redhat I can't reproduce the problem - it uses nplural from the PO if available, otherwise the value specified in the HLocale (if any), otherwise the default. 

Is it possible you forgot that the first first text flow was singular, not plural?

**Comment by djansen-redhat** [comment]
:white_check_mark: Tested

**Pull request merged**